ml-agents-envs/mlagents/envs/environment.py

```diff
@@ -279,156 +251,145 @@ def executable_launcher(self, file_name, docker_training, no_graphics, args):
                     shell=True,
                 )
 
-    def __str__(self):
-        return """Unity Academy name: {0}""".format(self._academy_name)
+    def _update_group_specs(self, output: UnityOutputProto) -> None:
+        init_output = output.rl_initialization_output
+        for brain_param in init_output.brain_parameters:
+            # Each BrainParameter in the rl_initialization_output should have at least one AgentInfo
+            # Get that agent, because we need some of its observations.
+            agent_infos = output.rl_output.agentInfos[brain_param.brain_name]
+            if agent_infos.value:
+                agent = agent_infos.value[0]
+                new_spec = agent_group_spec_from_proto(brain_param, agent)
+                self._env_specs[brain_param.brain_name] = new_spec
+                logger.info(f"Connected new brain:\n{brain_param.brain_name}")
 
-    def reset(self) -> AllBrainInfo:
+    def _update_state(self, output: UnityRLOutputProto) -> None:
         """
-        Sends a signal to reset the unity environment.
-        :return: AllBrainInfo  : A data structure corresponding to the initial reset state of the environment.
+        Collects experience information from all external brains in environment at current step.
         """
+        for brain_name in self._env_specs.keys():
+            if brain_name in output.agentInfos:
+                agent_info_list = output.agentInfos[brain_name].value
+                self._env_state[brain_name] = batched_step_result_from_proto(
+                    agent_info_list, self._env_specs[brain_name]
+                )
+            else:
+                self._env_state[brain_name] = BatchedStepResult.empty(
+                    self._env_specs[brain_name]
+                )
+        self._parse_side_channel_message(self.side_channels, output.side_channel)
+
+    @staticmethod
+    def _empty_action(spec: AgentGroupSpec, n_agents: int) -> np.array:
```

Comment:
Should this be a method on AgentGroupSpec instead?

Reply:
YES !